llama_index/vector_stores/simple.py

```diff
@@ -129,22 +156,34 @@ def query(
         **kwargs: Any,
     ) -> VectorStoreQueryResult:
         """Get nodes for response."""
-        if query.filters is not None:
+        # Prevent metadata filtering on stores that were persisted without metadata.
+        if (
+            query.filters is not None
+            and self._data.embedding_dict
+            and not self._data.metadata_dict
+        ):
             raise ValueError(
-                "Metadata filters not implemented for SimpleVectorStore yet."
+                "Cannot filter on metadata for stores that were persisted without metadata. "
+                "Please rebuild the store with metadata to enable filtering."
             )
+        # Prefilter nodes based on the query filter and node ID restrictions.
+        query_filter_fn = _build_metadata_filter_fn(
+            lambda node_id: self._data.metadata_dict[node_id], query.filters
+        )
 
-        # TODO: consolidate with get_query_text_embedding_similarities
-        items = self._data.embedding_dict.items()
-
-        if query.node_ids:
+        if query.node_ids is not None:
             available_ids = set(query.node_ids)
-
-            node_ids = [t[0] for t in items if t[0] in available_ids]
-            embeddings = [t[1] for t in items if t[0] in available_ids]
+            node_filter_fn = lambda node_id: node_id in available_ids
         else:
-            node_ids = [t[0] for t in items]
-            embeddings = [t[1] for t in items]
+            node_filter_fn = lambda _: True
+
+        node_ids = []
+        embeddings = []
+        # TODO: consolidate with get_query_text_embedding_similarities
+        for node_id, embedding in self._data.embedding_dict.items():
```

Comment:
Yea, still feels kind of gross to iterate over this one-by-one (I know we did this before too), but not sure I see faster way at the moment